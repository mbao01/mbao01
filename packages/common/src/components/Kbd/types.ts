import { type VariantProps } from "../../libs";
import { getKbdClasses } from "./constants";

export type KbdProps = React.HTMLAttributes<HTMLElement> & VariantProps<typeof getKbdClasses>;
