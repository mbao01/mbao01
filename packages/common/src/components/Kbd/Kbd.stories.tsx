import type { Meta, StoryObj } from "@storybook/react";
import { Kbd } from "./Kbd";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Kbd,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Kbd>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    children: "K",
  },
};

export const TinyKbd: Story = {
  args: {
    size: "xs",
    children: "Tiny",
  },
};

export const CompoundKbd: Story = {
  args: {},
  render: () => {
    return (
      <>
        Press <Kbd>⌘</Kbd> <Kbd>C</Kbd> or <Kbd>⌃</Kbd> <Kbd>C</Kbd> to cancel.
      </>
    );
  },
};

export const PrimaryKbd: Story = {
  args: {
    variant: "primary",
    children: "K",
  },
};

export const SuccessKbd: Story = {
  args: {
    variant: "success",
    children: "K",
  },
};

export const OutlineKbd: Story = {
  args: {
    outline: true,
    children: "K",
  },
};

export const OutlinePrimaryKbd: Story = {
  args: {
    outline: true,
    variant: "primary",
    children: "K",
  },
};

export const OutlineErrorKbd: Story = {
  args: {
    outline: true,
    variant: "error",
    children: "K",
  },
};
