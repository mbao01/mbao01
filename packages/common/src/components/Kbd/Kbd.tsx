import type { KbdProps } from "./types";
import { cn } from "../../utilities";
import { getKbdClasses } from "./constants";

export const Kbd = ({ outline, variant, size, className, ...props }: KbdProps) => (
  <kbd className={cn(getKbdClasses({ outline, variant, size }), className)} {...props} />
);
