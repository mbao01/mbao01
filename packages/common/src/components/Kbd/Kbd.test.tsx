import { render, screen } from "@testing-library/react";
import { Kbd } from "./Kbd";

describe("Kbd", () => {
  it("renders a basic Kbd", () => {
    const { asFragment } = render(<Kbd>K</Kbd>);

    expect(screen.getByText("K")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    ["has outline", true],
    ["has no outline", false],
  ] as const)("%s", (description, outline) => {
    const title = `Kbd ${description}`;
    const { asFragment } = render(
      <span data-testid="content">
        {title}:{" "}
        <Kbd outline={outline} variant="neutral">
          S
        </Kbd>
      </span>
    );

    expect(screen.getByTestId("content")).toHaveTextContent(`${title}: S`);
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    { size: "xs", name: "tiny" },
    { size: "sm", name: "small" },
    { size: "md", name: "base" },
    { size: "lg", name: "large" },
  ] as const)("has $name ($size) button", ({ size, name }) => {
    const title = `Kbd ${name}`;
    const { asFragment } = render(
      <span data-testid="content">
        {title}:{" "}
        <Kbd size={size} variant="neutral">
          S
        </Kbd>
      </span>
    );

    expect(screen.getByTestId("content")).toHaveTextContent(`${title}: S`);
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    "neutral",
    "primary",
    "secondary",
    "accent",
    "ghost",
    "info",
    "success",
    "warning",
    "error",
  ] as const)("has %s variant", (variant) => {
    const title = `Kbd ${variant}`;
    const { asFragment } = render(
      <span data-testid="content">
        {title}: <Kbd variant={variant}>B</Kbd>
      </span>
    );

    expect(screen.getByTestId("content")).toHaveTextContent(`${title}: B`);
    expect(asFragment()).toMatchSnapshot();
  });
});
