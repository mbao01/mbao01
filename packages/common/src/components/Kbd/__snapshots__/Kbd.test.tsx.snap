// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Kbd > has 'base' ('md') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd base: 
    <kbd
      class="kbd bg-neutral text-neutral-content border-neutral kbd-md"
    >
      S
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has 'large' ('lg') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd large: 
    <kbd
      class="kbd bg-neutral text-neutral-content border-neutral kbd-lg"
    >
      S
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has 'small' ('sm') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd small: 
    <kbd
      class="kbd bg-neutral text-neutral-content border-neutral kbd-sm"
    >
      S
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has 'tiny' ('xs') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd tiny: 
    <kbd
      class="kbd bg-neutral text-neutral-content border-neutral kbd-xs"
    >
      S
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has accent variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd accent: 
    <kbd
      class="kbd bg-accent text-accent-content border-accent"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has error variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd error: 
    <kbd
      class="kbd bg-error text-error-content border-error"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has ghost variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd ghost: 
    <kbd
      class="kbd bg-ghost text-ghost-content border-transparent"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has info variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd info: 
    <kbd
      class="kbd bg-info text-info-content border-info"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has neutral variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd neutral: 
    <kbd
      class="kbd bg-neutral text-neutral-content border-neutral"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has no outline 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd has no outline: 
    <kbd
      class="kbd bg-neutral text-neutral-content border-neutral"
    >
      S
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has outline 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd has outline: 
    <kbd
      class="kbd border-neutral bg-default text-neutral"
    >
      S
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has primary variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd primary: 
    <kbd
      class="kbd bg-primary text-primary-content border-primary"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has secondary variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd secondary: 
    <kbd
      class="kbd bg-secondary text-secondary-content border-secondary"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has success variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd success: 
    <kbd
      class="kbd bg-success text-success-content border-success"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > has warning variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Kbd warning: 
    <kbd
      class="kbd bg-warning text-warning-content border-warning"
    >
      B
    </kbd>
  </span>
</DocumentFragment>
`;

exports[`Kbd > renders a basic Kbd 1`] = `
<DocumentFragment>
  <kbd
    class="kbd"
  >
    K
  </kbd>
</DocumentFragment>
`;
